### PR TITLE
Configurable minimum characters before autocomplete user searches

### PR DESCRIPTION
**Pull request by VicDeo**
## Description
allow limit user/group search by a certain number of characters

## Related Issue
#30313 

## Motivation and Context
For large user lists searching for the first two characters can be very slow to return.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by VicDeo** [comment]
@PVince81 according to your spec from https://github.com/owncloud/core/issues/30313#issuecomment-368892658
>Backend code UserManager and GroupManager refuse to work if the search pattern given to any find* and search* methods is shorter than the minimum. This saves us for modifying backend for any other endpoints.

I don't see a better place to block a search pattern length. As for decorators: it's possible to extend User|Group manager with LimitedUserManager | LimitedGroupManager class and overload the methods in question. But it looks too nerdy to me :eyeglasses:

**Comment by PVince81** [comment]
But adding `isSearchable` on the `IUserManager` interface also feels wrong, especially considering that it's not really a capability with that name but a utility function.

@DeepDiver1975 @butonic @jvillafanez any other fancy ideas ?

**Comment by PVince81** [comment]
@VicDeo sounds good. You can even invent new utility services in the server container that suit our purposes.

**Comment by jvillafanez** [comment]
Maybe it's a good time to think about error handling in the user / group manager... returning an empty result just because you don't want to search with less than 4 chars feels bad because any user could think that there isn't any user, not that you're refusing to search with such little information.
Throwing an exception would be better, but it will likely need a lot of changes.

The problem is that the user / group manager doesn't specify any exception that could be thrown, so I don't expect any error handling to be in place, not even the default "catch all" exception handling with the "something went wrong" message.

For the specific check, it depends on how we want this to progress. If there are no plans to add similar functionality, adding the functions in the same class is fine if we consider it as implementation detail (and make those functions private). On the other hand, if we want to make restrictions based on the user type (maybe we want to allow the admin to do a less restrictive search), or any other, we'll likely need to move the functions to an specialized class.

**Comment by PVince81** [comment]
> returning an empty result just because you don't want to search with less than 4 chars feels bad because any user could think that there isn't any user, not that you're refusing to search with such little information

The other alternative is finding all the current HTTP API endpoints that we have and add the limit there. This includes the sharing autocomplete, custom groups and potentially others (gallery app old share dialog)

File: lib/public/Util/UserSearch.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ */
+
+
+namespace OCP\Util;
+use OCP\IConfig;
+
+/**
+ * This class provides functions to search users
+ *
+ * @since 10.1.0
```
Comment (PVince81) [inline lib/public/Util/UserSearch.php]: make that 10.0.8

**Review by PVince81 — APPROVED** [review]
Fair enough, it looks like it's the best compromise :+1: 

see comment

**Comment by VicDeo** [comment]
Comment addressed.
Fixed tests, rebased, squashed.

**Comment by PVince81** [comment]
follow up for missed test failure: https://github.com/owncloud/core/pull/30992

File: lib/public/Util/UserSearch.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+		return $trimmed === '' || strlen($trimmed) >= $this->getSearchMinLength();
+	}
+
+	/**
+	 * Get minimal allowed size to search users
+	 *
+	 * @return mixed
+	 */
+	public function getSearchMinLength() {
+		return $this->config->getSystemValue('user.search_min_length', 4);
```
[the remark this answers is not in the document]
Comment (VicDeo) [reply lib/public/Util/UserSearch.php]: @PVince81 we decided allow zero length for backward compatibility reasons.

**Comment by PVince81** [comment]
@VicDeo can you backport with both PRs ?

**Comment by VicDeo** [comment]
@PVince81 Sure. Do you mean 'both PRs as single PR'?

**Comment by VicDeo** [comment]
Stable10: https://github.com/owncloud/core/pull/30994

**Pull request merged**